Optimize indexing for the autogenerated ID append-only case

s1monw opened the pull request:
If elasticsearch controls the ID values as well as the documents
version we can optimize the code that adds / appends the documents
to the index. Essentially we an skip the version lookup for all
documents unless the same document is delivered more than once.

On the lucene level we can simply call IndexWriter#addDocument instead
of #updateDocument but on the Engine level we need to ensure that we deoptimize
the case once we see the same document more than once.

This is done as follows:

1. Mark every request with a timestamp. This is done once on the first node that
receives a request and is fixed for this request. This can be even the
machine local time (see why later). The important part is that retry
requests will have the same value as the original one.

2. In the engine we make sure we keep the highest seen time stamp of "retry" requests.
This is updated while the retry request has its doc id lock. Call this `maxUnsafeAutoIdTimestamp`

3. When the engine runs an "optimized" request comes, it compares it's timestamp with the
current `maxUnsafeAutoIdTimestamp` (but doesn't update it). If the the request
timestamp is higher it is safe to execute it as optimized (no retry request with the same
timestamp has been run before). If not we fall back to "non-optimzed" mode and run the request as a retry one
and update the `maxUnsafeAutoIdTimestamp` unless it's been updated already to a higher value

Relates to #19813

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -412,19 +421,77 @@ public boolean index(Index index) {
+                        if (highestDeOptimzeAddDocumentTimestamp.get() >= index.timestamp()) {
+                            break;
+                        }
+                    } while(highestDeOptimzeAddDocumentTimestamp.compareAndSet(deOptimizeTimestamp, index.timestamp()) == false);
```
we need to recapture deOptimizeTimestamp in the loop no?

jpountz commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -115,10 +118,16 @@
     private final AtomicInteger throttleRequestCount = new AtomicInteger();
     private final EngineConfig.OpenMode openMode;
     private final AtomicBoolean allowCommits = new AtomicBoolean(true);
+    private final AtomicLong highestDeOptimzeAddDocumentTimestamp = new AtomicLong(0);
```
missing `i` in `optimize`

jpountz commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -412,19 +421,77 @@ public boolean index(Index index) {
+                    useAddDocument = deOptimizeTimestamp < index.timestamp();
+                }
+                skipVersionCheck = useAddDocument == false; // we skip the version check if we can't use the optimization and just
+                // add update the document
```
remove "add" in the comment?

jpountz commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -412,19 +421,77 @@ public boolean index(Index index) {
             final long expectedVersion = index.version();
-            if (checkVersionConflict(index, currentVersion, expectedVersion, deleted)) return false;
+            if (skipVersionCheck == false) {
+                if(checkVersionConflict(index, currentVersion, expectedVersion, deleted)) {
```
this version conflict check should always be fast, so maybe we could remove `skipVersionCheck` entirely?

s1monw replied:
we can't otherwise we might run into version conflicts on retries if the original request make it in first and that would fail the document

mikemccand commented:
I tested performance on indexing geonames (8.6 M docs) on a fast 8 core box on a fast SSD with plenty of RAM (64 GB, vs 4 GB for JVM heap):

```
before:
 Total docs/sec: 44526.0
 Total docs/sec: 45460.2

after:
 Total docs/sec: 53650.4
 Total docs/sec: 53639.6
```

~20% speedup, nice :)

tlrx commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -919,6 +920,14 @@ protected int estimatedSizeInBytes() {
+            return isAutogeneratedID;
+        }
+
+        public boolean isCanHaveDuplicates() {
```
`public boolean canHaveDuplicates()` ?

tlrx commented on core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -89,7 +89,16 @@
-    private TranslogConfig translogConfig;
+    /**
+     * Configures an index to optimize documents with auto generated ids for append only. If this setting is updated to <code>false</code>
+     * it might not take affect immediately. Depending on the engine implementation it's anti-viral until an engine is restarted or the index is
```
affect -> effect

tlrx commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+             *  - is preserved in the transaction log
+             *  - and is assigned before we start to index / replicate
+             * NOTE: it's not important for this timestamp to be consistent across nodes etc. it's just a number that is in the common
+             * case increasing and can be used in the failure case when we retry and resent documents to establic a happens before relationship.
```
establic -> establish

tlrx commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+             *  while both docs are in in flight, we disconnect on one node, reconnect and send doc A again
+             *  - now doc A' has timestamp 10, autoGeneratedID = true, canHaveDuplicates = true
+             *
+             *  if A' arrives on the shard first we update maxUnoptimzeAddDocumentTimestamp to 10 and use update document. All subsequent
```
maxUnoptimzeAddDocumentTimestamp -> highestDeOptimizeAddDocumentTimestamp

tlrx commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+             *
+             *  if A' arrives on the shard first we update maxUnoptimzeAddDocumentTimestamp to 10 and use update document. All subsequent
+             *  documents that arrive (A and B) will also use updateDocument since their timestamps are less than maxUnoptimzeAddDocumentTimestamp.
+             *  While this is not stricly needed for doc B it is just much simpler to implement will just de-optimize some doc.
```
will -> while ?

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -589,10 +594,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedID == false;
+            autoGeneratedID = true;
```
can we explicitly set can have duplicates to false here (and default it true)? or at least assert it is false? (I prefer the first)

s1monw replied:
I dont' think we should do this to be honest I added assertions along the way to ensure it's only true if `autoGeneratedID` is true too also it's documented. We can switch so something like:

```
enum IDSource {
  External, AutoGenerated, RetryAutoGenerated
}
`` 
to have the different states clear, WDYT?

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -667,4 +676,30 @@ public String toString() {
+        if (autoGeneratedID) {
+            return canHaveDuplicates;
+        }
+        return false;
```
Nit: I think it's safe to return true here - although this is only relevant for the auto gen ID case it is better to indicate that a non-autogenedId request can always have duplicates.

bleskes replied:
Alternatively we can do what we do in the Engine.Index and just return the current value of `canHaveDuplicates` - I think is this the most straight forward.

s1monw replied:
this is very tricky and has caused some bugs along the way. I added assertions and I want to keep it the way it is. lots of tests had it wrong

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -89,7 +89,16 @@
-    private TranslogConfig translogConfig;
+    /**
+     * Configures an index to optimize documents with auto generated ids for append only. If this setting is updated to <code>false</code>
+     * it might not take effect immediately. Depending on the engine implementation it's anti-viral until an engine is restarted or the\
```
can you clarify what you mean by anti-viral and engine implementation?

s1monw replied:
sure

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -115,10 +119,18 @@
         openMode = engineConfig.getOpenMode();
+        if (engineConfig.getOptimizeAutoGeneratedIds() == false
+            || engineConfig.getIndexSettings().getIndexVersionCreated().before(Version.V_5_0_0_alpha6)) {
+            // no optimization for pre 5.0.0.alpha6 since translog might not have all information needed
```
what are we missing here? we have to do a full cluster restart, clearing any in flight request. This means all concurrency is between new requests?

s1monw replied:
old translogs won't have the info it it's autogenerated ID etc and it's important to do so

bleskes replied:
Sorry, I forget to respond to this earlier, but ops coming from the old translog won't have the time stamp and therefore will not be optimized. They can also not compete with new concurrent incoming request because they will have a different id (and we flush at the end of primary recovery, but we shouldn't rely on this).

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+             */
+            final boolean skipVersionCheck;
+            if (index.isAutoGeneratedID()
+                && index.version() == Versions.MATCH_ANY
```
Anything else for the version setting is not supported for auto gen id's. I think we change these to asserts here and hard enforce these in the IndexRequest when we auto gen the id / validate it?

bleskes replied:
At second thought, for replicas, the versionType will `EXTERNAL` here, so we can't assert on that either.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
         try (Releasable ignored = acquireLock(index.uid())) {
             lastWriteNanos = index.startTime();
-            final long currentVersion;
+            final boolean useAddDocument;
```
Nit: I think it will be easier to understand if we can call this `canHaveDuplicates` instead. It has more implications than just using AddDocument. Also, the boolean doesn't directly control the lucene code but rather indirectly through the `currentVersion` variable.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+                } else {
+                    useAddDocument = deOptimizeTimestamp < index.timestamp();
+                }
+                skipVersionCheck = useAddDocument == false; // we skip the version check if we can't use the optimization and just
```
I'm not sure this is correct. If the optimization is turned off because it's unsafe (i.e., useAddDocument is). We can do two things:

1) check ourselves by loading versions from the index or memory map and see if we already have the document. If we do, with don't re-index but respond with "OK".
2) don't check the current versions but rather index a new document (with version set to 1 etc) to lucene using updateDocument.

I'm in favor of two, allowing use to nuke the version map.

For clarity, right now the code does the normal thing which means loading versions and either reindexing with a version = 2 on the primary or rejecting the indexing operation on the replica.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1222,6 +1293,12 @@ public void onSettingsChanged() {
         // config().isEnableGcDeletes() or config.getGcDeletesInMillis() may have changed:
         maybePruneDeletedTombstones();
+        if (engineConfig.getOptimizeAutoGeneratedIds() == false) {
+            // this is an anti-viral settings you can only opt out for the entire index
```
I think I understand what your are saying - there is no way to re-enable this on a live engine - but it will be good to explicitly say that rather than "anti-viral" (if I get you correctly :))

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/TranslogRecoveryPerformer.java:
```diff
@@ -147,14 +147,18 @@ private void maybeAddMappingUpdate(String type, Mapping update, String docId, bo
                     Translog.Index index = (Translog.Index) operation;
+                    // we set canHaveDuplicates to true all the time such that we de-optimze the translog case and ensure that all
+                    // autoGeneratedID docs that are coming from the primary are updated correctly.
+                    final boolean canHaveDuplicates = index.isAutoGeneratedID();
```
another example where this is confusing - just use true?

bleskes commented on core/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -852,6 +854,11 @@ public Index(StreamInput in) throws IOException {
             this.ttl = in.readLong();
             this.versionType = VersionType.fromValue(in.readByte());
             assert versionType.validateVersionForWrites(this.version);
+            if (format >= FORMAT_AUTOGENERATED_IDS) {
```
Not part of your change, but I do wonder how that `assert format == SERIALIZATION_FORMAT : "format was: " + format;` assertion actually work with our BWC tests...

s1monw replied:
apparently we don't have any BWC tests that have translogs?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
+
+    public void testDoubleDelivery() throws IOException {
+        final ParsedDocument doc = testParsedDocument("test#1", "1", "test", null, 100, -1, testDocumentWithTextField(), new BytesArray("{}".getBytes(Charset.defaultCharset())), null);
+        Engine.Index operation = new Engine.Index(newUid("test#1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, Engine.Operation.Origin.PRIMARY, System.nanoTime(), true, false);
```
can we (randomly) use replica's operations?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -997,7 +1066,9 @@ private Releasable acquireLock(Term uid) {
         try (final Searcher searcher = acquireSearcher("load_version")) {
-            return Versions.loadVersion(searcher.reader(), uid);
+            long version = Versions.loadVersion(searcher.reader(), uid);
+            assert incrementLookupMetrics(version);
```
why do we only care about index based version lookups and not memory based lookups?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), 10);
+            assertEquals(1, topDocs.totalHits);
+        }
+        operation = new Engine.Index(newUid("test#1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, Engine.Operation.Origin.PRIMARY, System.nanoTime(), true, false);
```
why do we do this twice? I don't think there is any scenario where operation can arrive twice with its can have duplicates set to `false`?

s1monw replied:
well i am just double checking that this work that's all even if our system doesn't do it

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
         assertEquals(numDocs, docStats.getCount());
     }
+
+    public void testDoubleDelivery() throws IOException {
```
I'm a bit confused by this test compare to the other two - what do we test here that the other don't and in what scenarios that can happen? I feel like I'm missing something

s1monw replied:
nothing it's additional ways to test it. Its mainly copied from the stuff we had in 1.x and I am happy to have this tested in multiple ways

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
+        }
+    }
+
+    public void testRetryWithAutogeneratedIdsAndWrongOrderWorksAndNoDuplicateDocs() throws IOException {
```
do we have a  variant where the original request arrives second at the primary but first at the replica (and the reverse)?

s1monw replied:
yeah see the testDoubleDelivery test

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
+                    } catch (InterruptedException e) {
+                        throw new AssertionError(e);
+                    }
+                    int i;
```
can we use `docIndex` instead of `i` which is used for the threads?

s1monw replied:
nitpicker

bleskes commented on core/src/test/java/org/elasticsearch/index/replication/IndexLevelReplicationTests.java:
```diff
@@ -28,4 +40,64 @@ public void testSimpleReplication() throws Exception {
+                    super.cleanFiles(totalTranslogOps, sourceMetaData);
+                    latch.countDown();
+                    try {
+                        latch.await();
```
I wonder we should remove the concurrency here and fix the time the document is indexed to this moment - either by using two latches or just calling `shards.appendDocs` here?

s1monw replied:
why? can we just let it be concurrent just like reality?

bleskes replied:
for reproducibility sake. I'm good with leaving as is, if you prefer it.

bleskes commented:
I spent some hours thinking about this and going through the PR and I (still) really like the idea and simplicity of it. Next to the comments I left, I have the following suggestions:

1) I rather not use the timestamp field - it is a currently an external feature (which is deprecated if I read the code right) and if someone sets it, it can screw up with our logic (which is tricky to say the least).
2) I would love to see the new "highest time stamp seen from retry requests" state of the engine added to the engine stats. I can see how this would help us debug future issues where node clocks are way of sync and some network hickup causes the engine to be hours in non-optimized mode. I think it will be very hard to debug otherwise. 
3) I would love it if we can push as much of the logic as possible into the engine. Right now the canHaveDuplicate property of `IndexRequest` has a relationship with autoGenId. I would prefer if we change IndexRequest to have simple flags like "autoGenId" and "isRetry" (or isReplay) and let the engine draw its conclusions from that. I think it make the logic all in one place, making a hairy change easier.
4) I think it will be good to have an independent integration test that indexes in append only mode concurrently (with relocations and potentially with network partitions) and check the end result is good. Thinking about the past issues we had with this optimization, it's very likely that all components behave as expected but things still go wrong in subtle ways when they work together.
5) Why are we closing the meta issue of #19813 with this? I think there is more things to do, like disabling real time get and remove the version map etc.?

I also have some ideas on alternative implementations for (1) , will reach out to discuss.

s1monw commented:
>1) I rather not use the timestamp field - it is a currently an external feature (which is deprecated if I read the code right) and if someone sets it, it can screw up with our logic (which is tricky to say the least).

done

>2) I would love to see the new "highest time stamp seen from retry requests" state of the engine added to the engine stats. I can see how this would help us debug future issues where node clocks are way of sync and some network hickup causes the engine to be hours in non-optimized mode. I think it will be very hard to debug otherwise. 

done

> 3) I would love it if we can push as much of the logic as possible into the engine. Right now the canHaveDuplicate property of IndexRequest has a relationship with autoGenId. I would prefer if we change IndexRequest to have simple flags like "autoGenId" and "isRetry" (or isReplay) and let the engine draw its conclusions from that. I think it make the logic all in one place, making a hairy change easier.

done

> 4) I think it will be good to have an independent integration test that indexes in append only mode concurrently (with relocations and potentially with network partitions) and check the end result is good. Thinking about the past issues we had with this optimization, it's very likely that all components behave as expected but things still go wrong in subtle ways when they work together.

I beefed up the BackgroundIndexer to get them all for free

> 5) Why are we closing the meta issue of #19813 with this? I think there is more things to do, like disabling real time get and remove the version map etc.?

I am going to change this to Relates to: for now

@bleskes I am ready for another review

jpountz commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -589,10 +600,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedTimestamp == -1;
+            autoGeneratedTimestamp = System.currentTimeMillis();
```
paranoia: should we do `Math.max(0, System.currentTimeMillis())` to protect against the case someone's clock is set before 1970? (It might work to consider the id non automatic, but maybe it's better to just use 0 as a timestamp in that case?)

bleskes replied:
The "problematic" edge case is where someone has their time way in the future. Note both cases are "safe" in the sense that no duplicates will occur. Have the time in the future will disable the optimization for indexing request coming from other nodes. Having it in the past will cause problems on for this specific one. To deal with the first one we added the current timestamp used by the engine to the segments stats. See [here](https://github.com/elastic/elasticsearch/issues/19813#issuecomment-243079678).

bleskes replied:
and with more thought I think I might have totally misunderstood you and you worried about the default of -1 being "not set".  I agree that in that case we should use max(0, time) here.

jpountz replied:
yep, that's it :)

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -152,6 +152,11 @@ public static OpType fromString(String sOpType) {
 
     private String pipeline;
 
+    private long autoGeneratedTimestamp = -1;
```
nit: can we make -1 a constant and doc it? It has proven useful in the past to be able to see usages of these constants.

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -216,6 +221,12 @@ public ActionRequestValidationException validate() {
+
+        if (id == null && (versionType == VersionType.INTERNAL &&
+            (version == Versions.MATCH_ANY || version == Versions.MATCH_DELETED)) == false) {
+            validationException = addValidationError("an id must be provided if version type or value are change", validationException);
```
nit: "are change" -> "are set".

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -667,4 +682,28 @@ public String toString() {
+
+
+    /**
+     * Returns <code>true</code> if this request has been send to a shard copy more than once.
```
send -> sent.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -704,6 +704,7 @@ void retry(Exception failure) {
                 return;
             }
             setPhase(task, "waiting_for_retry");
+            request.onRetry();
```
can we add checks in the unit tests of this class (TransportReplicationActionTests) to make sure this is called?

s1monw replied:
I did my very best, can you check again

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -919,6 +921,23 @@ protected int estimatedSizeInBytes() {
+        }
+
+        /**
+         * Returns <code>true</code> iff this index requests has been retried on the coordinating node and can therefor be delivered
```
can we add a note about this being true for recovery (from the local translog or remote)

s1monw replied:
I don't think it's important since we handle it in the engine right even if it's not set

bleskes replied:
agreed. My comment was about the java docs - they say `iff`, which is strictly speaking not true entirely true

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
+                case PRIMARY:
+                    boolean value = (index.version() == Versions.MATCH_DELETED && index.versionType() == VersionType.INTERNAL)
+                        || (index.version() == Versions.MATCH_ANY && index.versionType() == VersionType.INTERNAL); // create request
+                    assert value; // we assert that since this was possible before and there might be some violations in the translog etc
```
I think the comment needs some fixing - old translogs will never get here, as the `getAutoGeneratedIdTimestamp` will be -1.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
         }
     }
 
+    private boolean canOptimizeAddDocument(Index index) {
```
This method is effectively  shortcut for `index.getAutoGeneratedIdTimestamp() >= 0`, with the exception of  `LOCAL_TRANSLOG_RECOVERY`. Since we always set is isRetry to true when replaying from the local translog, I'm not sure we need this? if we don't I would just convert to assertions and remove it.

s1monw replied:
I made all of the asserts I think we should still keep it as it contains it well

bleskes replied:
OK.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
+                }
+                versionValue = null;
+                currentVersion = Versions.NOT_FOUND;
+                deleted = index.version() == Versions.MATCH_DELETED;
```
I think this is a mistake. The previous code set delete to true if the doc wasn't found. We should do the same here.

bleskes replied:
PS. The delete naming is confusing (my fault), but see the java docs on `Versions.MATCH_DELETED`. I suspect we have not test that auto indexes with opType set to create. How about forcing that any index request with a custom opType must have an id? This will also simplify the assertions we have here in InternalEngine as we won't need to take `MATCH_DELETED` into account all the time.

s1monw replied:
it fails all tests all over the place if we do this - I think this is correct

s1monw replied:
we can only update this IFF it was deleted since it's a create case. so we have to set deleted to true in this cases otherwise we get a version conflict, makes sense?

s1monw replied:
nevermind that entire create business is obsolete here I am going to not optimize in that case

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
+                index.setCreated(true);
+                index(index, indexWriter);
+            } else {
+                update(index, versionValue, indexWriter);
```
there is a problem here with the index.setCreated flag when ends up being false on retry for auto gen id docs. I think we can move `index.setCreated` out of this if and do `index.setCreated(deleted)` just before the if (as a side effect, it will mean we don't need to pass `versionValue` to `update()` and can reduce the scope of that variable.

s1monw replied:
I need to think about this more - this has just be added recently

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -996,6 +1076,7 @@ private Releasable acquireLock(Term uid) {
     }
 
     private long loadCurrentVersionFromIndex(Term uid) throws IOException {
+        assert incrementIndexVersionLookup();
```
don't we want to increment this for in memory version lookups too?

s1monw replied:
we do there is a different one called `incrementVersionLookup` that is incremented in both situations

bleskes replied:
sorry, misread the name.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -89,7 +89,15 @@
-    private TranslogConfig translogConfig;
+    /**
+     * Configures an index to optimize documents with auto generated ids for append only. If this setting is updated to <code>false</code>
+     * it might not take effect immediately. Depending on the engine implementation a change to this setting won't be reflected until the
```
question: I think going from true to false is effective immediately, it's going from false to true that's problematic, right?

s1monw replied:
yes

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:
```diff
@@ -292,6 +306,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.byteSizeField(Fields.INDEX_WRITER_MEMORY_IN_BYTES, Fields.INDEX_WRITER_MEMORY, indexWriterMemoryInBytes);
         builder.byteSizeField(Fields.VERSION_MAP_MEMORY_IN_BYTES, Fields.VERSION_MAP_MEMORY, versionMapMemoryInBytes);
         builder.byteSizeField(Fields.FIXED_BIT_SET_MEMORY_IN_BYTES, Fields.FIXED_BIT_SET, bitsetMemoryInBytes);
+        builder.field(Fields.MAX_UNSAFE_AUTO_ID_TIMESTAMP, maxUnsafeAutoIdTimestamp);
```
nit: can we use `timeValueField` and have the same naming conventions? (in_millis at the end when we use a timestamp)

s1monw replied:
I don't think we should do this - it's an impl detail and we might go and use something else ie. a sequence id later that will mess things up?

bleskes replied:
I think we should inline with what it is today. It's indeed an implementation detail of the engine so I will be 100% ok with changing the stats if we change the implementation on the engine level.

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/TranslogRecoveryPerformer.java:
```diff
@@ -147,14 +147,17 @@ private void maybeAddMappingUpdate(String type, Mapping update, String docId, bo
-                        index.version(), index.versionType().versionTypeForReplicationAndRecovery(), origin);
-                    maybeAddMappingUpdate(engineIndex.type(), engineIndex.parsedDoc().dynamicMappingsUpdate(), engineIndex.id(), allowMappingUpdates);
+                        index.version(), index.versionType().versionTypeForReplicationAndRecovery(), origin, index.getAutoGeneratedIdTimestamp(), true);
+                    maybeAddMappingUpdate(engineIndex.type(), engineIndex.parsedDoc().dynamicMappingsUpdate(), index.id(), allowMappingUpdates);
```
nit: can we go back to engineIndex.id()?

bleskes commented on core/src/main/java/org/elasticsearch/index/translog/Translog.java:
```diff
@@ -828,8 +828,11 @@ public Source(BytesReference source, String routing, String parent, long timesta
 
     public static class Index implements Operation {
-        public static final int SERIALIZATION_FORMAT = 6; // since 2.0-beta1 and 1.1
+        public static final int FORMAT_2x = 6; // since 2.0-beta1 and 1.1
```
👍

bleskes commented on core/src/test/java/org/elasticsearch/index/store/ExceptionRetryIT.java:
```diff
@@ -137,5 +153,16 @@ public void sendRequest(DiscoveryNode node, long requestId, String action, Trans
+            }
+        }
+        assertTrue("exception must have been thrown otherwise setup is broken", exceptionThrown.get());
+        assertTrue("maxUnsafeAutoIdTimestamp must be > than 0 we have at least one retry", maxUnsafeAutoIdTimestamp > -1);
```
👍

bleskes commented on core/src/test/java/org/elasticsearch/indices/recovery/IndexPrimaryRelocationIT.java:
```diff
@@ -87,5 +93,11 @@ public void run() {
         finished.set(true);
         indexingThread.join();
+        refresh("test");
+        assertEquals(numAutoGenDocs.get(), client().prepareSearch("test").get().getHits().totalHits());
```
nit: use `assertHitCount`?

bleskes commented:
Thx @s1monw. I went through it again and made mostly cosmetic comments. Looking good!

s1monw commented:
@bleskes I addressed all your comments except of [this](https://github.com/elastic/elasticsearch/pull/20211#discussion_r76957771) I also wonder if you can take another look at the stuff I added in `TransportReplicationActionTests`

I think we are getting close though

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -216,6 +227,11 @@ public ActionRequestValidationException validate() {
                             id.getBytes(StandardCharsets.UTF_8).length, validationException);
         }
+
+        if (id == null && (versionType == VersionType.INTERNAL && version == Versions.MATCH_ANY) == false) {
```
can we put an extra validation that's explicit for the opType check - i.e., `if (opType() != OpType.INDEX && id == 1) addValidationError("an id is required for a " + opType() + " operation")` (preferably next the current ` if (opType() == OpType.CREATE) { ` line). I think people will be confused by this message if they use create and we should also catch the delete case, while at it.

bleskes commented on core/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -589,10 +605,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedTimestamp == -1;
+            autoGeneratedTimestamp = Math.min(0, System.currentTimeMillis()); // extra paranoia
```
I think we want a Max here?

s1monw replied:
LOL yes

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -407,30 +420,108 @@ public void index(Index index) {
-            if (versionValue == null) {
-                currentVersion = loadCurrentVersionFromIndex(index.uid());
-                deleted = currentVersion == Versions.NOT_FOUND;
+            final VersionValue versionValue;
```
this can now go into the else scope of `if (canOptimizeAddDocument(index)) {`, starting at line 500. It's not used elsewhere.

bleskes commented:
@s1monw it looks great to me, with the exception of this [comment](https://github.com/elastic/elasticsearch/pull/20211#discussion_r77016280). Feel free to reject or except the rest. They are minor. The TransportReplicationActionTests look good too.

s1monw commented:
@bleskes thx so much for the in-depth reviews, this would have been impossible without that! I will wait for CI to run and push tomorrow morning

mikemccand commented:
I re-tested performance gain on latest PR, still ~20% speedup!  Nice:

```
before:

  Total docs/sec: 44299.1
  Total docs/sec: 43192.2  

after:

  Total docs/sec: 53321.2
  Total docs/sec: 52980.7
```

bleskes commented:
@mikemccand w00t. I presume this was without replicas?

mikemccand commented:
@bleskes yes: single node with defaults, except 4 GB heap.
